<a href="https://colab.research.google.com/github/shcho11/04.IBM_AIEngineering_Pytorch/blob/main/Pytorch_LinearRegression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1>Linear Regression: Prediction</h1>

- 함께 참고할 만한 wikidocs : 'PyTorch로 시작하는 딥 러닝 입문-03.선형회귀(Linear Regression)' https://wikidocs.net/53545


<h2>Objective</h2><ul><li> How to make the prediction for multiple inputs.</li><li> How to use linear class to build more complex models.</li><li> How to build a custom module.</li></ul> 


<h2>Table of Contents</h2>

*In this lab, we will  review how to make a prediction in several different ways by using PyTorch.*

- Prediction
- Class Linear
- Build Custom Modules


<h2 id="Prediction">Prediction</h2>


In [2]:
import torch

Let us create the following expressions:


$b=-1,w=2$

$\hat{y}=-1+2x$


First, define the parameters:


In [3]:
# Define w = 2 and b = -1 for y = wx + b

w = torch.tensor(2.0, requires_grad = True)
b = torch.tensor(-1.0, requires_grad = True)

# requires_grad=True 로 설정하면 , 텐서의 모든 연산에 대해 추적을 시작한다. (디폴트는 False이다.)

Then, define the function <code>forward(x, w, b)</code> makes the prediction:


In [4]:
# Function forward(x) for prediction

def forward(x):
    yhat = w * x + b
    return yhat

Let's make the following prediction at <i>x = 1</i>


$\hat{y}=-1+2x$

$\hat{y}=-1+2(1)$


In [5]:
# Predict y = 2x - 1 at x = 1

x = torch.tensor([[1.0]])
yhat = forward(x)
print("The prediction: ", yhat)

The prediction:  tensor([[1.]], grad_fn=<AddBackward0>)


Now, let us try to make the prediction for multiple inputs:


<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0110EN/notebook_images%20/chapter2/2.1.2.png" width="500" alt="Linear Regression Multiple Input Samples" />


Let us construct the <code>x</code> tensor first. Check the shape of <code>x</code>.


In [6]:
# Create x Tensor and check the shape of x tensor

x = torch.tensor([[1.0], [2.0]])
print("The shape of x: ", x.shape)

The shape of x:  torch.Size([2, 1])


Now make the prediction:


In [7]:
# Make the prediction of y = 2x - 1 at x = [1, 2]

yhat = forward(x)
print("The prediction: ", yhat)

The prediction:  tensor([[1.],
        [3.]], grad_fn=<AddBackward0>)


In [8]:
# Practice: Make a prediction of y = 2x - 1 at x = [[1.0], [2.0], [3.0]]

x = torch.tensor([[1.0], [2.0], [3.0]])

yhat = forward(x)

print("The Prediction : ", yhat)

The Prediction :  tensor([[1.],
        [3.],
        [5.]], grad_fn=<AddBackward0>)


<h2 id="Linear">Class Linear</h2>


The linear class can be used to make a prediction. We can also use the linear class to build more complex models. Let's import the module:
- torch.nn이 무엇인가요 ? https://tutorials.pytorch.kr/beginner/nn_tutorial.html?highlight=linear 
-  파이토치에서 이미 구현되어져 제공되고 있는 함수들을 불러오는 것으로 더 쉽게 선형 회귀 모델을 구현해보겠음

In [9]:
# Import Class Linear
from torch.nn import Linear

Set the random seed because the parameters are randomly initialized:


In [10]:
# Set random seed
# # 현재 실습하고 있는 파이썬 코드를 재실행해도 다음에도 같은 결과가 나오도록 랜덤 시드(random seed)를 줍니다.
torch.manual_seed(1)

Let us create the linear object by using the constructor. The parameters are randomly created. Let us print out to see what <i>w</i> and <i>b</i>. The parameters of an <code>torch.nn.Module</code> model are contained in the model’s parameters accessed with <code>lr.parameters()</code>:


In [11]:
# Create Linear Regression Model, and print out the parameters

lr = Linear(in_features=1, out_features=1, bias=True)
print("Parameters w and b: ", list(lr.parameters()))

# in_features   : size of each input sample
# out_features  : size of each output sample
# bias          : If set to False, the layer will not learn an additive bias. Default: True

Parameters w and b:  [Parameter containing:
tensor([[0.5153]], requires_grad=True), Parameter containing:
tensor([-0.4414], requires_grad=True)]


This is equivalent to the following expression:


$b=-0.44, w=0.5153$

$\hat{y}=-0.44+0.5153x$


A method  <code>state_dict()</code> Returns a Python dictionary object corresponding to the layers of each parameter  tensor.

- state_dict 는 간단히 말해 각 계층을 매개변수 텐서로 매핑되는 Python 사전(dict) 객체입니다.
- 참고 : https://tutorials.pytorch.kr/recipes/recipes/what_is_state_dict.html

In [12]:
print("Python dictionary: ",lr.state_dict())
print("keys: ",lr.state_dict().keys())
print("values: ",lr.state_dict().values())

Python dictionary:  OrderedDict([('weight', tensor([[0.5153]])), ('bias', tensor([-0.4414]))])
keys:  odict_keys(['weight', 'bias'])
values:  odict_values([tensor([[0.5153]]), tensor([-0.4414])])


The keys correspond to the name of the attributes and the values correspond to the parameter value.


In [13]:
print("weight:",lr.weight)
print("bias:",lr.bias)

weight: Parameter containing:
tensor([[0.5153]], requires_grad=True)
bias: Parameter containing:
tensor([-0.4414], requires_grad=True)


Now let us make a single prediction at <i>x = \[\[1.0]]</i>.


In [14]:
# Make the prediction at x = [[1.0]]

x = torch.tensor([[1.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.0739]], grad_fn=<AddmmBackward0>)


Similarly, you can make multiple predictions:


<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0110EN/notebook_images%20/chapter2/2.1.2vector_function.png" width="500" alt="Linear Class Sample with Multiple Inputs" />


Use model <code>lr(x)</code> to predict the result.


In [15]:
# Create the prediction using linear model

x = torch.tensor([[1.0], [2.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.0739],
        [0.5891]], grad_fn=<AddmmBackward0>)


<h3>Practice</h3>


Make a prediction of the following <code>x</code> tensor using the linear regression model <code>lr</code>.


In [16]:
# Practice: Use the linear regression model object lr to make the prediction.

x = torch.tensor([[1.0],[2.0],[3.0]])
yhat = lr(x)
print("The Prediction : ", yhat)

The Prediction :  tensor([[0.0739],
        [0.5891],
        [1.1044]], grad_fn=<AddmmBackward0>)


<h2 id="Cust">Build Custom Modules</h2>


Now, let's build a custom module. We can make more complex models by using this method later on.


First, import the following library.


In [17]:
# Library for this section

from torch import nn

Now, let us define the class:


In [18]:
# Customize Linear Regression Class

class LR(nn.Module):
    
    # Constructor
    def __init__(self, input_size, output_size):
        
        # Inherit from parent
        super(LR, self).__init__()
        self.linear = nn.Linear(input_size, output_size)
    
    # Prediction function
    def forward(self, x):
        out = self.linear(x)
        return out

Create an object by using the constructor. Print out the parameters we get and the model.


In [19]:
# Create the linear regression model. Print out the parameters.

lr = LR(1, 1)
print("The parameters: ", list(lr.parameters()))
print("Linear model: ", lr.linear)

The parameters:  [Parameter containing:
tensor([[-0.1939]], requires_grad=True), Parameter containing:
tensor([0.4694], requires_grad=True)]
Linear model:  Linear(in_features=1, out_features=1, bias=True)


Let us try to make a prediction of a single input sample.


In [20]:
# Try our customize linear regression model with single input

x = torch.tensor([[1.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.2755]], grad_fn=<AddmmBackward0>)


Now, let us try another example with multiple samples.


In [21]:
# Try our customize linear regression model with multiple input

x = torch.tensor([[1.0], [2.0]])
yhat = lr(x)
print("The prediction: ", yhat)

The prediction:  tensor([[0.2755],
        [0.0816]], grad_fn=<AddmmBackward0>)


the parameters are also stored in an ordered dictionary :


In [22]:
print("Python dictionary: ", lr.state_dict())
print("keys: ",lr.state_dict().keys())
print("values: ",lr.state_dict().values())

Python dictionary:  OrderedDict([('linear.weight', tensor([[-0.1939]])), ('linear.bias', tensor([0.4694]))])
keys:  odict_keys(['linear.weight', 'linear.bias'])
values:  odict_values([tensor([[-0.1939]]), tensor([0.4694])])


<h3>Practice</h3>


Create an object <code>lr1</code> from the class we created before and make a prediction by using the following tensor:


In [23]:
# Practice: Use the LR class to create a model and make a prediction of the following tensor.

x = torch.tensor([[1.0], [2.0], [3.0]])

lr1 = LR(1,1)
yhat = lr(x)
print("The Prediction : ", yhat)

The Prediction :  tensor([[ 0.2755],
        [ 0.0816],
        [-0.1122]], grad_fn=<AddmmBackward0>)
